fix: Improve text and fix mistakes in blog post (#12)
savonet/blog: _posts/2024-10-26-a-faster-liquidsoap-part-deux.md

## The labyrinth of memory allocation

We've had a lot of discussions about how memory is allocated in Liquidsoap over the years.
[One of the most recent ones](https://github.com/savonet/liquidsoap/issues/4058) was a great opportunity to realize that
it is, in fact, very tricky to actually _measure_ how much memory a process is consuming.

Here's a rough rundown of how memory is allocated that might explain that a bit.

### Memory pages

Memory is organized into fixed-size chunks called _pages_. On Linux, a page is typically 4 KB, while on Apple Silicon,
it is usually 16 KB.

Each process is associated with a collection of memory pages that are assigned to it by the OS kernel. When the process
requests memory allocation, the OS looks for available memory on the currently allocated pages and, if there is not
enough, allocates more pages for the process.

Once a page has had some of its memory used by the process, it is considered _dirty_. Before a page can be reclaimed,
the process must free all the memory on that page.

![Paged Memory Illustration](https://github.com/user-attachments/assets/f617f88d-4a7c-4214-bb9d-6d57cfaf3bd5)

> When a process requests `1024` bytes of memory, it often results in `4k` bytes being allocated instead. Similarly, if a

Reviewer: I don't think that this is accurate:
```suggestion
When a process requests `1024` bytes of memory, it may result in `4k` bytes being allocated instead. Similarly, if a
```

process releases `1024` bytes of memory, its effective memory usage may not decrease. This can happen if the page
containing that memory still has other sections in use.

### Shared memory

Not all memory used by a process belongs to it! Typically, when a process loads a dynamic library, the binary code from

> that library is potentially shared across all processes loading from the same library.

Reviewer: ```suggestion
that library is potentially shared across all processes using it.
```

Technically, this means that the binary code from the library is loaded in _shared memory pages_. The kernel can

> associate these pages with multiple processes that utilize the same shared library.

Reviewer: ```suggestion
associate these pages with multiple processes that load the same shared library.
```

  /opt/homebrew/opt/jpeg-turbo/lib/libjpeg.8.dylib (compatibility version 8.0.0, current version 8.3.2)
  /opt/homebrew/opt/giflib/lib/libgif.dylib (compatibility version 0.0.0, current version 7.2.0)
  /opt/homebrew/opt/jack/lib/libjack.0.1.0.dylib (compatibility version 0.1.0, current version 1.9.22)
  /opt/homebrew/opt/libao/lib/libao.4.dylib (compatibility version 6.0.0, current version 6.1.0)
  /usr/lib/libcurl.4.dylib (compatibility version 7.0.0, current version 9.0.0)
```

Yeah, that's **a lot!**

The more shared libraries your process has, the more shared memory it will use. However, all the processes using the
same libraries also reuse this memory.

![Shared Memory Illustration](https://github.com/user-attachments/assets/34bf149f-b198-42b4-a4ab-731df23a709a)

This means that if you have `10` Liquidsoap processes, each using `100MB` of memory, including shared memory, the total
memory used on the system is not actually `1000MB` because a significant portion is shared!

### Shared memory and docker processes?

Before proceeding, it's important to note that for memory sharing from dynamic libraries between programs, the system

> must recognize that the programs are indeed utilizing the same shared libraries.

Reviewer: Let's be more technical for clarity:
```suggestion
must recognize that the programs are indeed loading the same shared libraries.
```

layer that is not entirely controllable by the application, typically when memory collection happens and how much and
how long memory is allocated before it can be freed.

This is even more important in the case of Liquidsoap because, by nature of its streaming loop, the application
constantly allocates memory under very short streaming cycles (typically the duration of a frame).

If memory collection occurs too frequently, CPU usage noticeably increases. Conversely, if it happens too infrequently,
memory usage rises.

We [explored these issues in a previous post](../2023-07-09-memory-management) and gave an example of parameters that
govern this trade-off. In the future, we would like to explore alternative memory allocation strategies to better
control short-term memory allocations inside the application.

### How to report memory usage

Now that we've seen different elements of memory usage, how can we report memory usage? It turns out that this is tricky
and OS-dependent!

In Linux, if you use `top` or a similar tool, the most common measure of memory consumption is `RSS` for
[_Resident set size_](https://en.wikipedia.org/wiki/Resident_set_size).

> While this number includes shared memory, it does not imply that if `10` processes each use `30MB` of RSS memory, the

Reviewer: The logical statement is:
* Make clear that RSS does include shared memory
* Therefore, process memory is not additive

```suggestion
One important thing about RSS memory usage is that it includes shared memory. So, if `10` processes each use `30MB` of RSS memory, the
```

> total memory used by the system is `300MB`.

Reviewer: ```suggestion
total memory used by the system is not going to be `300MB`.
```

Moreover, if a program like Liquidsoap depends on many shared libraries, the total reported `RSS` memory may be quite
large; however, it might not accurately reflect the actual memory the application occupies.

Likewise, in [the previously mentioned issue](https://github.com/savonet/liquidsoap/issues/4058), it took us a while to
realize that the tool our user was using to report the process's memory usage from their virtualized environment was
also including the page cache, leading to the wrong observations regarding the memory effectively allocated by the
application itself.

In the [ocaml-mem_usage](https://github.com/savonet/ocaml-mem_usage) module, we implemented several OS-specific methods

> to measure the application's memory use and separate shared memory from private memory.

Reviewer: I'm not sure that this is more clear.

Author: This fragment is literally the same, except the word order.
```diff
- the memory used by the application
+ the application's memory use
```

Author: Thank you, I have undone this change.

Reviewer: Thanks. Yes, it is the same meaning but I think that, for technical documentation being explicit at the expense of being repetitive/less idiomatic is always better. So, "the memory used by the application" is always more clear, especially for non-native reader, than "the application's memory" which requires understanding the english genitive, a construct specific to that language.
